docs/src/tutorial.md

```diff
@@ -0,0 +1,197 @@
+@variable(model, x[i=1:2], start=x0[i])
+@NLobjective(model, Min, (x[1] - 1)^2 + 100 * (x[2] - x[1]^2)^2)
+optimize!(model)
+```
+
+Another example, using a constrained problem
+
+```@example ex1
+n = 10
+x0 = ones(n)
+x0[1:2:end] .= -1.2
+nlp = ADNLPModel(x -> sum((x[i] - 1)^2 + 100 * (x[i+1] - x[i]^2)^2 for i = 1:n-1), x0,
+                 c=x -> [3 * x[k+1]^3 + 2 * x[k+2] - 5 + sin(x[k+1] - x[k+2]) * sin(x[k+1] + x[k+2]) +
+                         4 * x[k+1] - x[k] * exp(x[k] - x[k+1]) - 3 for k = 1:n-2],
+                 lcon=zeros(n-2), ucon=zeros(n-2))
+stats = ipopt(nlp, print_level=0)
+print(stats)
+```
+
+## Output
+
+The output of `ipopt` is a `GenericExecutionStats` from `SolverTools`. It contains a few basic information of the solver.
+In addition to the built-in fields of `GenericExecutionStats`, we also store in
+`solver_specific` the following fields:
+
+- `multipliers_con`: Constraints multipliers;
+- `multipliers_L`: Variables lower-bound multipliers;
+- `multipliers_U`: Variables upper-bound multipliers;
+- `internal_msg`: Detailed Ipopt output message.
+
+```@example ex1
+stats.solver_specific[:internal_msg]
+```
+
+## Manual input
+
+This is an example of manually inputting the problem and its derivatives. For this,
+we're creating an NLPModel, and we need to define the following API functions:
+- `obj(nlp, x)`: objective
+- `grad!(nlp, x, g)`: gradient
+- `cons!(nlp, x, c)`: constraints, if any
+- `rows, cols = jac_structure(nlp)`: structure of the Jacobian, if constrained;
+- `jac_coord!(nlp, x, rows, cols, vals)`: Jacobian values (don't access `rows` and
+  `cols`, as Ipopt doesn't actually pass them;
+- `hess_structure(nlp)`: structure of the lower half of the Hessian of the
+  Lagrangian;
+- `hess_coord!(nlp, x, rows, cols, vals; obj_weight=1.0, y=[])`: Hessian of the
+  Lagrangian, where `obj_weight` is the weight assigned to the objective, and `y` is the
+  multipliers vector.
+
+Let's implement a logistic regression model. We consider the model ``h(\beta; x) = (1 + e^{-\beta^Tx})^{-1}``, and the loss function
+```math
+\ell(\beta) = -\sum_{i = 1}^m y_i \ln h(\beta; x_i) + (1 - y_i) \ln(1 - h(\beta; x_i))
+```
+with regularization ``\lambda \|\beta\|^2 / 2``.
+
+```@example ex3
+using DataFrames, LinearAlgebra, NLPModels, NLPModelsIpopt, Random
+
+mutable struct LogisticRegression <: AbstractNLPModel
+  X :: Matrix
+  y :: Vector
+  λ :: Real
+  meta :: NLPModelMeta # required by AbstractNLPModel
+  counters :: Counters # required by AbstractNLPModel
+end
+
+function LogisticRegression(X, y, λ = 0.0)
+  m, n = size(X)
+  meta = NLPModelMeta(n, name="LogisticRegression", nnzh=div(n * (n+1), 2) + n) # nnzh is the length of the coordinates vectors
+  return LogisticRegression(X, y, λ, meta, Counters())
+end
+
+function NLPModels.obj(nlp :: LogisticRegression, β::AbstractVector)
+  hβ = 1 ./ (1 .+ exp.(-nlp.X * β))
+  -sum(nlp.y .* log.(hβ .+ 1e-8) .+ (1 .- nlp.y) .* log.(1 .- hβ .+ 1e-8)) + nlp.λ * dot(β, β) / 2
```

Comment:
This doesn't look very nice because of the line break. Could we either put it all on one long line, or use `hβ += ...` on line 2 (and perhaps 3)?

Reply:
The second line is supposed to be the return value, I'll add a return.